### DISCUSSION TITLE
Offset camera/nozzle

### DISCUSSION THREAD

**Discussion by Fruitfarmer1** [geezacoleman/OpenWeedLocator#37] [General]
Hi. I found a little conversation on AgOpenGPS about this project, and had to have a crack.
I have just received my rasp pi's, cameras and other bit and pieces. I want to build two units and mount one each side of my tractor to spray out underneath my orchard trees.
I think this will be an ideal use for OWL as the tree rows are always maintained as a fallow and I only want to control one nozzle per OWL unit, and ground speed will be about 4km/hr.
I will need to mount the units offset with the camera looking in under the tree from the side, I think this will reduce the units ability to identify the green weeds to some extent but I'm hopeful that with some tinkering I'll get a good result.
Any thoughts or recommendations on how to achieve a good result?
Cheers.

**Comment by geezacoleman** [comment]
That's great! Glad to hear you're building a system - I'm much more a cropping systems person so my orchard agronomy/context is a bit limited, but hopefully a few points below are useful. 

Only controlling a single nozzle simplifies things a lot, you'll just need to change the [default number of nozzles to one](https://github.com/geezacoleman/OpenWeedLocator/blob/main/greenonbrown.py#L139) in greenonbrown.py:
```
139                 nozzleNum=1,
```
Or just set the `nozzleNum=1` variable under line 464, as you would with any of the others. I just noticed we haven't included it in the other variables there - I'll update that now. Haven't used this with single nozzles before, but just tested the code on my computer then and it seems to run fine. You'll also be able to do away with the 4-channel relay board and replace with the cheaper single relay. The code from lines [313 to 332](https://github.com/geezacoleman/OpenWeedLocator/blob/main/greenonbrown.py#L313-L332) could be simplified a lot to make it run faster with only one nozzle, but it works fine as is. 

What sort of angle are you thinking of mounting the OWL on and what height? The depth of field on those cameras is ok, but if you have quite a difference between the close objects and the distant ones you might run into issues with focus. If the camera is looking too far off into the distance/bright horizon you might end up with quite a few false positives. They could be tuned out with the HSV thresholding though. Might be worth initially just experimenting with different angles/heights and working out which is the best balance.

The offset between the unit and the nozzle is important too. The GPS-based dynamic offset/delay hasn't been implemented yet, so you'll just need to work out what the distance between the camera and the nozzle is and then work out how long it will need to wait before spraying.

Not sure how dense the canopy will be, but would be worth considering your lighting too - when we tested on overcast days or in really dense shadows, the blur can creep up a lot. 4kmh will limit the impact of that but worth considering if you're not happy with initial results. Would also recommend using the latest software with exposure_compensation=-6 to help reduce the difference between dark shandows/bright highlights if it's more dappled light.

If you're operating at 4kmh, you might be able to bump the resolution up a touch too. Might not double but could help with some of the weeds further from the camera.

Would love to see photos of the rig once you've got it set up!

**Reply by Fruitfarmer1** [reply]
Thanks for the reply and the suggested code changes. The nozzle is about 700mm from the ground and the sprayed out strip is 2.8m wide, so 1.5m each side of tree.
So the tractor is driving up to 2.2m wide grass inter-row spraying a 1.5m wide strip each side. The nozzles are Hypro XT boomless nozzles. These nozzles are used instead of small booms on the back of ATV's etc. and councils sometimes use them to spray out on the sides of roads on the front of Utes and small trucks etc.
I think the field of view may be a challenge, but I guess the best thing to do is build it and try it.
Will let you know how it turns out.

**Reply by geezacoleman** [reply]
With the current set up we work off ~62 degree horizontal field of view from the camera. Using a bit of trigonometry you should be able to work out the best angle/height to mount it at to get you in the ballpark at least before putting it all together.

One thing with those nozzles is to check how long they take from turning on, to developing the correct spray width. For solenoid options, you may be better off using something like [this too](https://www.ebay.com.au/itm/125260984983?chn=ps&_ul=AU&_trkparms=ispr%3D1&amdata=enc%3A19onwzyjDRFe-2MzXIMpz0g55&norover=1&mkevt=1&mkrid=705-139619-5960-0&mkcid=2&itemid=125260984983&targetid=&device=c&mktype=pla&googleloc=1000286&poi=&campaignid=15984321586&mkgroupid=&rlsatarget=&abcId=9300814&merchantid=494515424&gclid=CjwKCAjws8yUBhA1EiwAi_tpEe13sn-SN614Ir7oXm4JjpEVQBojg-G0xlajrPtFU4B1KfX9xA8k3BoC_zcQAvD_BwE) so you can mount it near/directly adjacent to the Hypro XTs to minimise how much it dribbles after shutoff:
![s-l500](https://user-images.githubusercontent.com/51358498/170857522-2dc9ec62-31c8-43c1-9d65-5cff1d9d723f.jpg)
 
They are a lot cheaper, but durability is possibly more questionable. 

As you mention, there are a lot of use cases for a setup like this so it will be great to see how it turns out and what we can change to make it work better.

**Comment by Fruitfarmer1** [comment]
Good suggestion, I currently use a Braglia high pressure solenoid manifold but I can easily mount them at the nozzle.They're a good solenoid. On 29 May 2022 5:10 pm, Guy Coleman ***@***.***> wrote:
With the current set up we work off ~62 degree horizontal field of view from the camera. Using a bit of trigonometry you should be able to work out the best angle/height to mount it at to get you in the ballpark at least before putting it all together.
One thing with those nozzles is to check how long they take from turning on, to developing the correct spray width. For solenoid options, you may be better off using something like this too so you can mount it near/directly adjacent to the Hypro XTs to minimise how much it dribbles after shutoff:

They are a lot cheaper, but durability is possibly more questionable.
As you mention, there are a lot of use cases for a setup like this so it will be great to see how it turns out and what we can change to make it work better.
